ci/script/functions.sh

```diff
@@ -9,14 +9,20 @@ SPECS_HAVE_RUN_FILE=specs.out
 MAINTENANCE_BRANCH=`cat maintenance-branch`
 
 # Don't allow rubygems to pollute what's loaded. Also, things boot faster
```

Comment:
Can we revert all these changes for now, it makes diffing harder and its easy to clean up post release.

Reply:
With an exception of `BRANCH_TO_CLONE` trick, we can. But we equally can just set `s.required_ruby_version = '>= 2.3.0'` and clean everything else after the release 😆 

Jokes aside, `rspec-mock` has undertaken [the smallest change, _just_  +360 −880](https://github.com/rspec/rspec-mocks/pull/1349) of all repos. Not much compared to [+45 -145 induced by this update](https://github.com/rspec/rspec-mocks/pull/1355).

All of the changes to `ci/script/functions.sh` and `ci/script/predicate_functions.sh` consist of removals and simple changes a-la `is_mri_2plus` -> `is_mri`. That can be skimmed through really quickly during the review (of the total `4-0-dev` -> `main` merge I suppose).

Except for the [`BRANCH_TO_CLONE` trick](https://github.com/rspec/rspec-core/commit/6c657734123171aab1811c7926c563bbd27b8013#diff-f686af268076eb6078295d47e8d97ab3df9486f0f22b900f4c045b5b9cc04fe7R22) that is in `rspec-*` repos' `4-0-dev`, but not on `main`, and not here, in `rspec-dev` anywhere. I've added it to this PR to avoid it being overwritten with `rake ci:update_files` next time we update `4-0-dev`.